### ISSUE TITLE
Thank you

### ISSUE THREAD

**Issue by giantbirdicus** [gabrielolympie/sglang-flashnext-sm120#1]
Thanks for posting this. I really appreciate it. Helped me out a lot. 
I was trying to get qwen3.8-flash-next ncfp4 running on sglang with patches myself unsuccessfully and came across your repo.

I'm using your work with a Dell precision 5820 W-2275 with 128GB (quad channel) DDR4 memory and the RTX 6000 Pro.

I haven't run any tests, just got it running and playing with it and looking at nvidia-smi metrics I track with prometheus and the litellm proxy log i pass this through and,

For single concurrency its as advertised, I feel it seems to go faster on my machine than your readme states.

however,

even though I limit my context length to 161k I seem to be getting VRAM memory thrashing when more than 2 concurrent requests for work like claude code (agentic coding)

for document research and editing it is fantastic not having to wait forever for qwen3.8-27b to finish thinking.

If I figure out my memory thrashing problem I'lls hare. I have a bit more ram to spare I might try using HiCache ram/ssd split to see if that band-aids over the issue.

Again thanks for sharing your work.

**Comment by giantbirdicus** [comment]
to help prevent random OOM errors while using I had to add

export SGLANG_PREFILL_KERNEL_MODE=1
export SGLANG_WARMUP_KERNELS=1

to server.sh then it become much more stable for me.

I haven't tried the hicache yet to see if it smooths out (what looks like memory thrashingto me) for concurrency of 4 yet.

**Comment by giantbirdicus** [comment]
I figured out my memory thrashing observation (...of the VRAM on the GPU).

1. in my .env file I had typed MEMFRAC wrong so it was using the default 0.85 in the script
2. I have lost the ability to do grade school arithmetic in my head and had expected 163K of context available for each of the 4 concurrent requests so my agents where able to send more context than available VRAM. When I checked the math for my machine/settings the max was actually ~153k available for each of the 4 concurrent request.

with those 2 fixes it feels like the memory thrashing is gone however i only have 900MB of vram left so likely for stability that 153K context should be more like ~ 140K and the MEMFRAC i set to 0.981 (pulled from looking at the scripts in the repo) probably needs to come down as i suspect i'll probably still get the odd OOM over the course of a few days.

This is awesome. feels like i have a fast frontier model running locally. 
I also think my quad channel ram is helpful because with 3 concurrent requests going along  i can see 650-690 tps, yes its variable this is just best i've seen.

anyway thanks again for sharing your work I really appreciate it.

**Comment by gabrielolympie** [comment]
Hey, thanks for the appreciation, much welcome :)

For speed, I guess the difference comes from the GPU, as I have a maxq it might power throttle sometime.

Agreed on concurrency, under load with 4 concurrent request I rarely get over 400tps, so a bit limiting for batched work, my guess is that the triton kernel Claude wrote were optimized for single stream, I might try to optimize it a bit for multi stream next weekend.

Otherwise, I'm mostly using it with pi agent (oh my pi flavour) for long running workflow, and it's great, I recommand, totally cross compatible with Claude code

**Comment by giantbirdicus** [comment]
Yes switching to pi is on my list, thanks for the recommendation i will try oh-my-pi.

so the attached image was 4 agents doing research based tasks, IMO its fast and stable. 

I am using HiCache with a cache split of 10GB CPU RAM and 24 GB SSD. I dropped down to max 150K token length

I still occasionally get a crash because no memory for a 'pong' - sorry I didn't copy the error - it just comes down to tweaking all the settings

I haven't figured out where I need MEMFRAC to be 0.9-0.95. I'm at 0.9 and its been stable for a while at 150K total context length and hicache helps smooth out TTFT for KV cache ejections on agentic concurrency.

I'm going to play with this ratio next I haven't added it yet but next crash I will:
  --mamba-full-memory-ratio 0.3

I too am using the RTX  6000 pro maxq. i set fan to 90% and depending on temp I'll drop down to max power of 290 and hardly notice a difference in performance.

below is a screen grab of 4 agents it was maintaining this speed for a while and I wasn't seeing the thrashing I saw initially. It must be the quad channel DDR4 helping out here for both the embeddings of the model and the ram hicache.

<img width="1689" height="1030" alt="Image" src="https://github.com/user-attachments/assets/b1b34409-7a4a-456c-97b4-33b53fd59463" />

**Comment by gabrielolympie** [comment]
Nice, gonna try that, i have a few pipelines that can use that concurrency ^^

**Comment by gabrielolympie** [comment]
Otherwise, below are my launch configs, i'm mostly using the 800k context version so far, and its stable even at 0.98 max mem limit,
Mean tps is about 230, with peaks up to 270 on function calling or object structuration

  qwen3.8-flash-next:
    backend: sglang
    served_model_name: pennyroyal
    env:
      MEMFRAC: "0.95"          # weights 81.5 GB + ~3.6 GB fp8 copies; < 0.92 leaves no room for KV
      KVDTYPE: fp8_e4m3        # fp8 KV (bf16 halves capacity); needs patches/0002
      SGLANG_SM120_LOWM_FP8_WEIGHT: "1"  # W8A16 fp8 copies of dense weights for the sm120 low-M Triton GEMM (commits 086a37f+284d7fe)
      SGLANG_SM120_LM_HEAD_FP8: "1"     # fp8 lm_head for decode logits (85e7da1); gates+GSM 5/5 pass
      LINEAR_BACKEND: flashinfer   # FlashInfer GDN on sm120; needs patches/0003
      SSM_DTYPE: bfloat16
      MAMBA_RADIX: extra_buffer
      SPEC: "1"                # native NEXTN MTP, 3 steps / 4 draft tokens
      # Relaxed acceptance (lossy at temp>0: sharpens toward high-prob tokens).
      # Ladder @temp0.6 C1: 1.0 lossless=179, 0.5=203, 0.3=231. Gates pass at 0.3.
      SPEC_ACCEPT_SINGLE: "0.3"
      SPEC_ACCEPT_ACC: "0.3"
      # FR-Spec: draft lm_head over a 64K hot-token subset (exact verify, only
      # draft proposals restricted). C1 201->219, C4 541->592. "none" disables.
      SPEC_TOKEN_MAP: /home/golympie/ai-toolbox/models/qwen38fn/hot_tokens_64k.pt
      CTX: "262144"            # full native window, zero speed cost (KV pool is VRAM-bound,
                               # ~406K fp8 tokens). 151K needles pass at all depths. For a
                               # ~800K KV pool instead: SGLANG_SM120_LOWM_FP8_WEIGHT=0 +
                               # MEMFRAC=0.98 + MAXREQ=4 + MAMBA_CACHE=24 (costs ~20% decode).
      # Concurrency 4 (user choice 2026-08-31). MAMBA_CACHE must be ~6x MAXREQ
      # for MTP or speculative CUDA graphs silently cap below MAXREQ.
      MAXREQ: "4"
      CUDAGRAPH_MAXBS: "4"
      MAMBA_CACHE: "24"
      GDN_MTP_CACHE_MODE: none # RecoverSSM + FlashInfer WY output-only MTP verify (patches/0001b); validated 2026-08-30

  qwen3.8-flash-next-800k:
    backend: sglang
    served_model_name: pennyroyal
    env:
      MEMFRAC: "0.98"
      CTX: "786432"
      MAXREQ: "2"
      CUDAGRAPH_MAXBS: "2"
      MAMBA_CACHE: "12"
      SGLANG_SM120_LOWM_FP8_WEIGHT: "0"
      KVDTYPE: fp8_e4m3
      LINEAR_BACKEND: flashinfer
      SSM_DTYPE: bfloat16
      MAMBA_RADIX: extra_buffer
      SPEC: "1"
      SPEC_ACCEPT_SINGLE: "0.3"
      SPEC_ACCEPT_ACC: "0.3"
      GDN_MTP_CACHE_MODE: none
      ROPE_OVERRIDE: '{"text_config":{"rope_parameters":{"mrope_interleaved":true,"mrope_section":[11,11,10],"rope_type":"yarn","rope_theta":10000000,"partial_rotary_factor":0.25,"factor":3.0,"original_max_position_embeddings":262144}}}'
